// enhanced-resolve/lib/dist/MainFieldPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var path = require("path");

var DescriptionFileUtils = require("./DescriptionFileUtils");
/** @typedef {import("./Resolver")} Resolver */

/** @typedef {import("./Resolver").JsonObject} JsonObject */

/** @typedef {import("./Resolver").ResolveRequest} ResolveRequest */

/** @typedef {import("./Resolver").ResolveStepHook} ResolveStepHook */

/** @typedef {{name: string|Array<string>, forceRelative: boolean}} MainFieldOptions */


var alreadyTriedMainField = Symbol("alreadyTriedMainField");

module.exports =
/*#__PURE__*/
function () {
  /**
   * @param {string | ResolveStepHook} source source
   * @param {MainFieldOptions} options options
   * @param {string | ResolveStepHook} target target
   */
  function MainFieldPlugin(source, options, target) {
    _classCallCheck(this, MainFieldPlugin);

    this.source = source;
    this.options = options;
    this.target = target;
  }
  /**
   * @param {Resolver} resolver the resolver
   * @returns {void}
   */


  _createClass(MainFieldPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var _this = this;

      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("MainFieldPlugin", function (request, resolveContext, callback) {
        if (request.path !== request.descriptionFileRoot ||
        /** @type {ResolveRequest & { [alreadyTriedMainField]?: string }} */
        request[alreadyTriedMainField] === request.descriptionFilePath || !request.descriptionFilePath) return callback();
        var filename = path.basename(request.descriptionFilePath);
        var mainModule =
        /** @type {string|null|undefined} */
        DescriptionFileUtils.getField(
        /** @type {JsonObject} */
        request.descriptionFileData, _this.options.name);

        if (!mainModule || typeof mainModule !== "string" || mainModule === "." || mainModule === "./") {
          return callback();
        }

        if (_this.options.forceRelative && !/^\.\.?\//.test(mainModule)) mainModule = "./" + mainModule;
        /** @type {ResolveRequest & { [alreadyTriedMainField]?: string }} */

        var obj = _objectSpread({}, request, _defineProperty({
          request: mainModule,
          module: false,
          directory: mainModule.endsWith("/")
        }, alreadyTriedMainField, request.descriptionFilePath));

        return resolver.doResolve(target, obj, "use " + mainModule + " from " + _this.options.name + " in " + filename, resolveContext, callback);
      });
    }
  }]);

  return MainFieldPlugin;
}();
