// enhanced-resolve/lib/dist/ModulesInHierarchicalDirectoriesPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var forEachBail = require("./forEachBail");

var getPaths = require("./getPaths");
/** @typedef {import("./Resolver")} Resolver */

/** @typedef {import("./Resolver").ResolveRequest} ResolveRequest */

/** @typedef {import("./Resolver").ResolveStepHook} ResolveStepHook */


module.exports =
/*#__PURE__*/
function () {
  /**
   * @param {string | ResolveStepHook} source source
   * @param {string | Array<string>} directories directories
   * @param {string | ResolveStepHook} target target
   */
  function ModulesInHierarchicalDirectoriesPlugin(source, directories, target) {
    _classCallCheck(this, ModulesInHierarchicalDirectoriesPlugin);

    this.source = source;
    this.directories =
    /** @type {Array<string>} */
    [].concat(directories);
    this.target = target;
  }
  /**
   * @param {Resolver} resolver the resolver
   * @returns {void}
   */


  _createClass(ModulesInHierarchicalDirectoriesPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var _this = this;

      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("ModulesInHierarchicalDirectoriesPlugin", function (request, resolveContext, callback) {
        var fs = resolver.fileSystem;
        var addrs = getPaths(
        /** @type {string} */
        request.path).paths.map(function (p) {
          return _this.directories.map(function (d) {
            return resolver.join(p, d);
          });
        }).reduce(function (array, p) {
          array.push.apply(array, p);
          return array;
        }, []);
        forEachBail(addrs,
        /**
         * @param {string} addr addr
         * @param {(err?: null|Error, result?: null|ResolveRequest) => void} callback callback
         * @returns {void}
         */
        function (addr, callback) {
          fs.stat(addr, function (err, stat) {
            if (!err && stat && stat.isDirectory()) {
              /** @type {ResolveRequest} */
              var obj = _objectSpread({}, request, {
                path: addr,
                request: "./" + request.request,
                module: false
              });

              var message = "looking for modules in " + addr;
              return resolver.doResolve(target, obj, message, resolveContext, callback);
            }

            if (resolveContext.log) resolveContext.log(addr + " doesn't exist or is not a directory");
            if (resolveContext.missingDependencies) resolveContext.missingDependencies.add(addr);
            return callback();
          });
        }, callback);
      });
    }
  }]);

  return ModulesInHierarchicalDirectoriesPlugin;
}();
