// enhanced-resolve/lib/dist/RootsPlugin.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Ivan Kopeykin @vankop
*/
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var forEachBail = require("./forEachBail");
/** @typedef {import("./Resolver")} Resolver */

/** @typedef {import("./Resolver").ResolveRequest} ResolveRequest */

/** @typedef {import("./Resolver").ResolveStepHook} ResolveStepHook */


var RootsPlugin =
/*#__PURE__*/
function () {
  /**
   * @param {string | ResolveStepHook} source source hook
   * @param {Set<string>} roots roots
   * @param {string | ResolveStepHook} target target hook
   */
  function RootsPlugin(source, roots, target) {
    _classCallCheck(this, RootsPlugin);

    this.roots = Array.from(roots);
    this.source = source;
    this.target = target;
  }
  /**
   * @param {Resolver} resolver the resolver
   * @returns {void}
   */


  _createClass(RootsPlugin, [{
    key: "apply",
    value: function apply(resolver) {
      var _this = this;

      var target = resolver.ensureHook(this.target);
      resolver.getHook(this.source).tapAsync("RootsPlugin", function (request, resolveContext, callback) {
        var req = request.request;
        if (!req) return callback();
        if (!req.startsWith("/")) return callback();
        forEachBail(_this.roots,
        /**
         * @param {string} root root
         * @param {(err?: null|Error, result?: null|ResolveRequest) => void} callback callback
         * @returns {void}
         */
        function (root, callback) {
          var path = resolver.join(root, req.slice(1));
          /** @type {ResolveRequest} */

          var obj = _objectSpread({}, request, {
            path: path,
            relativePath: request.relativePath && path
          });

          resolver.doResolve(target, obj, "root path ".concat(root), resolveContext, callback);
        }, callback);
      });
    }
  }]);

  return RootsPlugin;
}();

module.exports = RootsPlugin;
