// rollup/dist/dist/rollup.dev.js
/*
  @license
	Rollup.js v4.41.1
	Sat, 24 May 2025 06:13:57 GMT - commit 7c469dc4eb8e1cb6def9fdc04581fdfce9975da3

	https://github.com/rollup/rollup

	Released under the MIT License.
*/
'use strict';

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

Object.defineProperty(exports, Symbol.toStringTag, {
  value: 'Module'
});

var rollup = require('./shared/rollup.js');

var parseAst_js = require('./shared/parseAst.js');

var fseventsImporter = require('./shared/fsevents-importer.js');

require('node:process');

require('node:path');

require('path');

require('./native.js');

require('node:perf_hooks');

require('node:fs/promises');

var WatchEmitter =
/*#__PURE__*/
function () {
  function WatchEmitter() {
    _classCallCheck(this, WatchEmitter);

    this.currentHandlers = Object.create(null);
    this.persistentHandlers = Object.create(null);
  } // Will be overwritten by Rollup


  _createClass(WatchEmitter, [{
    key: "close",
    value: function close() {
      return regeneratorRuntime.async(function close$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
            case "end":
              return _context.stop();
          }
        }
      });
    }
  }, {
    key: "emit",
    value: function emit(event) {
      for (var _len = arguments.length, parameters = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        parameters[_key - 1] = arguments[_key];
      }

      return Promise.all([].concat(_toConsumableArray(this.getCurrentHandlers(event)), _toConsumableArray(this.getPersistentHandlers(event))).map(function (handler) {
        return handler.apply(void 0, parameters);
      }));
    }
  }, {
    key: "off",
    value: function off(event, listener) {
      var listeners = this.persistentHandlers[event];

      if (listeners) {
        // A hack stolen from "mitt": ">>> 0" does not change numbers >= 0, but -1
        // (which would remove the last array element if used unchanged) is turned
        // into max_int, which is outside the array and does not change anything.
        listeners.splice(listeners.indexOf(listener) >>> 0, 1);
      }

      return this;
    }
  }, {
    key: "on",
    value: function on(event, listener) {
      this.getPersistentHandlers(event).push(listener);
      return this;
    }
  }, {
    key: "onCurrentRun",
    value: function onCurrentRun(event, listener) {
      this.getCurrentHandlers(event).push(listener);
      return this;
    }
  }, {
    key: "once",
    value: function once(event, listener) {
      var _this = this;

      var selfRemovingListener = function selfRemovingListener() {
        _this.off(event, selfRemovingListener);

        return listener.apply(void 0, arguments);
      };

      this.on(event, selfRemovingListener);
      return this;
    }
  }, {
    key: "removeAllListeners",
    value: function removeAllListeners() {
      this.removeListenersForCurrentRun();
      this.persistentHandlers = Object.create(null);
      return this;
    }
  }, {
    key: "removeListenersForCurrentRun",
    value: function removeListenersForCurrentRun() {
      this.currentHandlers = Object.create(null);
      return this;
    }
  }, {
    key: "getCurrentHandlers",
    value: function getCurrentHandlers(event) {
      return this.currentHandlers[event] || (this.currentHandlers[event] = []);
    }
  }, {
    key: "getPersistentHandlers",
    value: function getPersistentHandlers(event) {
      return this.persistentHandlers[event] || (this.persistentHandlers[event] = []);
    }
  }]);

  return WatchEmitter;
}();

function watch(configs) {
  var emitter = new WatchEmitter();
  watchInternal(configs, emitter)["catch"](function (error) {
    rollup.handleError(error);
  });
  return emitter;
}

function withTrailingSlash(path) {
  if (path[path.length - 1] !== '/') {
    return "".concat(path, "/");
  }

  return path;
}

function checkWatchConfig(config) {
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = config[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var item = _step.value;

      if (item.input && item.output) {
        var input = typeof item.input === 'string' ? rollup.ensureArray(item.input) : item.input;
        var output = rollup.ensureArray(item.output);

        var _loop = function _loop(index) {
          var inputPath = input[index];
          var subPath = output.find(function (o) {
            if (!o.dir || typeof inputPath !== 'string') {
              return false;
            }

            var _outPath = withTrailingSlash(o.dir);

            var _inputPath = withTrailingSlash(inputPath);

            return _inputPath.startsWith(_outPath);
          });

          if (subPath) {
            parseAst_js.error(parseAst_js.logInvalidOption('watch', parseAst_js.URL_WATCH, "the input \"".concat(inputPath, "\" is a subpath of the output \"").concat(subPath.dir, "\"")));
          }
        };

        for (var index in input) {
          _loop(index);
        }
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"] != null) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
}

function watchInternal(configs, emitter) {
  var optionsList, watchOptionsList, _ref, Watcher;

  return regeneratorRuntime.async(function watchInternal$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return regeneratorRuntime.awrap(Promise.all(rollup.ensureArray(configs).map(function (config) {
            return rollup.mergeOptions(config, true);
          })));

        case 2:
          optionsList = _context2.sent;
          watchOptionsList = optionsList.filter(function (config) {
            return config.watch !== false;
          });

          if (!(watchOptionsList.length === 0)) {
            _context2.next = 6;
            break;
          }

          return _context2.abrupt("return", parseAst_js.error(parseAst_js.logInvalidOption('watch', parseAst_js.URL_WATCH, 'there must be at least one config where "watch" is not set to "false"')));

        case 6:
          checkWatchConfig(watchOptionsList);
          _context2.next = 9;
          return regeneratorRuntime.awrap(fseventsImporter.loadFsEvents());

        case 9:
          _context2.next = 11;
          return regeneratorRuntime.awrap(Promise.resolve().then(function () {
            return require('./shared/watch.js');
          }));

        case 11:
          _ref = _context2.sent;
          Watcher = _ref.Watcher;
          new Watcher(watchOptionsList, emitter);

        case 14:
        case "end":
          return _context2.stop();
      }
    }
  });
}

exports.VERSION = rollup.version;
exports.defineConfig = rollup.defineConfig;
exports.rollup = rollup.rollup;
exports.watch = watch;
