// fdir/dist/api/functions/dist/push-file.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.build = build;

var pushFileFilterAndCount = function pushFileFilterAndCount(filename, _paths, counts, filters) {
  if (filters.every(function (filter) {
    return filter(filename, false);
  })) counts.files++;
};

var pushFileFilter = function pushFileFilter(filename, paths, _counts, filters) {
  if (filters.every(function (filter) {
    return filter(filename, false);
  })) paths.push(filename);
};

var pushFileCount = function pushFileCount(_filename, _paths, counts, _filters) {
  counts.files++;
};

var pushFile = function pushFile(filename, paths) {
  paths.push(filename);
};

var empty = function empty() {};

function build(options) {
  var excludeFiles = options.excludeFiles,
      filters = options.filters,
      onlyCounts = options.onlyCounts;
  if (excludeFiles) return empty;

  if (filters && filters.length) {
    return onlyCounts ? pushFileFilterAndCount : pushFileFilter;
  } else if (onlyCounts) {
    return pushFileCount;
  } else {
    return pushFile;
  }
}
