// fdir/dist/api/functions/dist/push-directory.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.build = build;

function pushDirectoryWithRelativePath(root) {
  return function (directoryPath, paths) {
    paths.push(directoryPath.substring(root.length) || ".");
  };
}

function pushDirectoryFilterWithRelativePath(root) {
  return function (directoryPath, paths, filters) {
    var relativePath = directoryPath.substring(root.length) || ".";

    if (filters.every(function (filter) {
      return filter(relativePath, true);
    })) {
      paths.push(relativePath);
    }
  };
}

var pushDirectory = function pushDirectory(directoryPath, paths) {
  paths.push(directoryPath || ".");
};

var pushDirectoryFilter = function pushDirectoryFilter(directoryPath, paths, filters) {
  var path = directoryPath || ".";

  if (filters.every(function (filter) {
    return filter(path, true);
  })) {
    paths.push(path);
  }
};

var empty = function empty() {};

function build(root, options) {
  var includeDirs = options.includeDirs,
      filters = options.filters,
      relativePaths = options.relativePaths;
  if (!includeDirs) return empty;
  if (relativePaths) return filters && filters.length ? pushDirectoryFilterWithRelativePath(root) : pushDirectoryWithRelativePath(root);
  return filters && filters.length ? pushDirectoryFilter : pushDirectory;
}
